### PR TITLE
1.0 Compatible Version of #313. Fixes `logp` and adds  `softmax` functionality

### PR DESCRIPTION
**Pull request by ekinakyurek**
I realized that current `logp` does not use cudnn function because of wrong condition. There was a mistake for dims==2 too. I replicated #313 for Julia 1.0 and all problems are solved.

### CODE REVIEW

File: src/loss.jl [hunk 1]
```diff
@@ -101,6 +85,56 @@ mutable structdef enum
 
 =#          
 
+
+"""
+    softmax(x;[dims])
```
Comment (CarloLucibello) [inline src/loss.jl]: softmax(x; dims =:, algo=1)

File: test/loss.jl [hunk 2]
```diff
@@ -28,5 +28,7 @@ include("header.jl")
         @test isapprox(nll(k, indices, dims=1), nll(a, indices, dims=1))
         @test isapprox(nll(k, indices, dims=2), nll(a, indices, dims=2))
     end
+    logistic(a[:],a[:])
+    bce(a[:],a[:])
```
Comment (CarloLucibello) [inline test/loss.jl]: this is not a test

File: src/loss.jl [hunk 3]
```diff
@@ -171,6 +205,31 @@ function nll(y,a::AbstractArray{<:Integer}; dims=1, average=true)
...
+    ε = eltype(x̂)(1e-12)
+    l = log.((1-ε) .+ exp.(-x .* x̂))
+    average ? mean(l) : sum(l)
+end
+
+"""
+
+    bce(scores,answers;average=true)
+
+Computes binary cross entropy given scores(predicted values) and answer labels.
```
Comment (CarloLucibello) [inline src/loss.jl]: space after scores

File: test/loss.jl (hunk 2 above)
Comment (CarloLucibello) [reply test/loss.jl]: also, logsoftmax should be tested

File: src/loss.jl [hunk 4]
```diff
@@ -101,6 +85,56 @@ mutable structdef enum
...
+"""
+const logsoftmax = logp
+
+function generic_softmax(x,algo::Int,fallback;dims=:)
+    if isa(value(x), KnetArray)
+        if dims==1
+            sz = size(x)
+            x = cudnnSoftmaxForward(reshape(x, (1,1,sz[1],:)), algo=algo)
+            reshape(x, sz)
+        elseif dims==2 && ndims==2
```
Comment (CarloLucibello) [inline src/loss.jl]: ndims -> ndims(x)

An error like this should be caught when tested on a gpu. If not, please add a test

**Comment by CarloLucibello** [comment]
There are some parts of #313  that I don't see here, and some tests also. I think
it is better to simply rebase #313, then merging this as it is. 
We can shrink this pr to just the addition of the bce and logistic losses.
Also, when building upon a stale PR, it would be nice to preserve the authorship of the first commits. It is
not hard to do, just a matter of branching for the stale branch.

File: src/loss.jl (hunk 3 above)
Comment (ekinakyurek) [reply src/loss.jl]: done

File: src/loss.jl (hunk 1 above)
Comment (ekinakyurek) [reply src/loss.jl]: I tried be consistent with master. We may go this interface.

File: src/loss.jl (hunk 4 above)
Comment (ekinakyurek) [reply src/loss.jl]: tests added, typo is corrected.

File: test/loss.jl (hunk 2 above)
Comment (ekinakyurek) [reply test/loss.jl]: logsoftmax is jus an alias. This will increase test time. Do we really need it ?

File: src/loss.jl (hunk 1 above)
Comment (ekinakyurek) [reply src/loss.jl]: Sorry, you can use this functionality, but I didn't expose it in the documentation.

**Pull request closed without merging**